### PR TITLE
Updated regex for more descriptive error messages

### LINKED ISSUE #161: [PE-D] Editing a module to an empty description is not accepted
I have the current module in my list:

I wish to modify the description to be empty, so I type the following command, and is given an error:

Screenshot of the UG:

As there is no mention that my description cannot be empty, may I safely assume that empty descriptions are allowed and this is a bug?

My purpose of editing a module to have an empty description is because I wish to remove the existing description.

-------------
Labels: `severity.Medium` `type.FeatureFlaw`
original: alanlowzies/ped#5

### PR DESCRIPTION
**Pull request by heekit73098**
Did some basic testing on the regex now, so far dont seem to spot any bugs yet, will continue to test.
Split EditParser and AddParser to its Module and Task counterparts as the regex is quite complicated, so will need to update the DG as well.

PPP and TA's comments will be updated on a separate PR

### CODE REVIEW

File: src/main/java/seedu/duke/parsers/AddTaskParser.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+     * (\s+-t\s+\"(?<estimatedWorkingTime>[^\"]+)\")?)   -- matches [-t "estimatedWorkingTime"] if present. Optional
+     *                                                   -- None of the above fields accept " as a valid character.
+     *
+     * (?<invalid>.*)                                    -- matches [invalid]
+     *                                                      Any other excess inputs
+     */
+
+    private static final String ADD_FORMAT = "(task\\s+\\\"(?<taskName>[^\\\"]+)\\\"(\\s+(-m\\s+(?<taskModule>\\w+)))?"
+            + "(\\s+-d\\s+\\\"(?<taskDescription>[^\\\"]+)\\\")?"
+            + "(\\s+-t\\s+\\\"(?<estimatedWorkingTime>[^\\\"]+)\\\")?)(?<invalid>.*)";
```
Comment (Ch40gRv1-Mu) [inline src/main/java/seedu/duke/parsers/AddTaskParser.java:47]: This fix the issue [#177](https://github.com/AY2122S2-CS2113T-T10-3/tp/issues/177)

File: src/main/java/seedu/duke/parsers/ModHappyParser.java [hunk 2]
```diff
@@ -73,13 +83,27 @@ private Parser getCommandParser(String commandWord) throws UnknownCommandExcepti
...
         case (ADD_COMMAND_WORD):
-            return new AddParser();
+            switch (parsedCommand.get(ARGUMENT).split(SPACE)[0]) {
+            case TASK:
+                return new AddTaskParser();
+            case MODULE:
+                return new AddModuleParser();
+            default:
+                throw new UnknownCommandException();
+            }
```
Comment (Ch40gRv1-Mu) [inline src/main/java/seedu/duke/parsers/ModHappyParser.java]: I suggest to put the logic to AddParser, and let AddParser to decide whether it's module or task to keep ModHappyParser in consistent logical level. (May let AddParser decide whether return AddTaskParser or AddModuleParser to ModHappyParser)

File: src/main/java/seedu/duke/parsers/ModHappyParser.java [hunk 3]
```diff
@@ -73,13 +83,27 @@ private Parser getCommandParser(String commandWord) throws UnknownCommandExcepti
...
         case (EDIT_COMMAND_WORD):
-            return new EditParser();
+            switch (parsedCommand.get(ARGUMENT).split(SPACE)[0]) {
+            case TASK:
+                return new EditTaskParser();
+            case MODULE:
+                return new EditModuleParser();
+            default:
+                throw new UnknownCommandException();
+            }
```
Comment (Ch40gRv1-Mu) [inline src/main/java/seedu/duke/parsers/ModHappyParser.java]: Same suggestion as in ModHappyParser

**Review by Ch40gRv1-Mu — CHANGES_REQUESTED** [review]
Good optimization, I leaves suggestion on the logic of ModHappyParser and EditParser. I also suggest to add some JUnit for the two command for(Add and edit) testing purpose, you may add it in the current PR, or I will add it later. This PR significantly fixed the issue [#177](https://github.com/AY2122S2-CS2113T-T10-3/tp/issues/177)

File: src/main/java/seedu/duke/util/StringConstants.java [hunk 4]
```diff
@@ -178,12 +178,13 @@
     /**
      * For exceptions.
      */
+    public static final String COMPULSORY_PARAMETERS = "compulsory parameters";
     public static final String ERROR_NO_SUCH_MODULE = "Sorry, no such module exists ._.";
     public static final String ERROR_NO_SUCH_TASK = "Sorry, no such task exists ._.";
     public static final String ERROR_PARSE_FAILED = "This parse failed 0_0";
-    public static final String ERROR_PARSE_INVALID_PARAM = "\nInvalid compulsory parameters. "
+    public static final String ERROR_PARSE_INVALID_PARAM = "\nInvalid %s. "
```
Comment (Yzkkk) [inline src/main/java/seedu/duke/util/StringConstants.java]: Do you think we should quote the user's invalid input just to be consistent?

File: src/main/java/seedu/duke/util/StringConstants.java [hunk 5]
```diff
@@ -193,12 +194,22 @@
...
+    public static final String ERROR_MISSING_TAG_OPERATION = "\nMissing operation."
+            + "\nPlease try again. Accepted commands are: add, del.";
+
     public static final String ERROR_INVALID_MODULE_GRADE = "\nInvalid module grade '%s'."
             + "\nPlease try again. Accepted module grades are: A+, A, A-, B+, B, B-, C+, C, D+, D, F, CS, CU, S, U.";
-    public static final String ERROR_INVALID_MODULE_CODE = "\nInvalid module code.\nPlease try again.";
-    public static final String ERROR_INVALID_NUMBER = "\nInvalid number format '%s'."
+    public static final String ERROR_MISSING_MODULE_GRADE = "\nMissing module grade."
+            + "\nPlease try again. Accepted module grades are: A+, A, A-, B+, B, B-, C+, C, D+, D, F, CS, CU, S, U.";
+    public static final String ERROR_INVALID_NUMBER = "\nInvalid number format for %s."
```
Comment (Yzkkk) [inline src/main/java/seedu/duke/util/StringConstants.java]: same question as above.

File: src/main/java/seedu/duke/util/StringConstants.java [hunk 6]
```diff
@@ -193,12 +194,22 @@
     public static final String ERROR_INVALID_FLAG = "\nInvalid flag '%s'."
             + "\nPlease check and try again. "
             + "\nYou may input 'help' followed by your command word to view the expected input format.";
+
+    public static final String ERROR_MISSING_FLAG = "\nMissing flag."
+            + "\nPlease check and try again. "
+            + "\nYou may input 'help' followed by your command word to view the expected input format.";
     public static final String ERROR_INVALID_TAG_OPERATION = "\nInvalid operation '%s'."
```
Comment (Yzkkk) [inline src/main/java/seedu/duke/util/StringConstants.java:202]: An example would be this, user's wrong input will be quoted

File: src/main/java/seedu/duke/util/StringConstants.java [hunk 7]
```diff
@@ -211,8 +222,10 @@
             + "View all available config settings with \"option\".";
     public static final String ERROR_MODULE_LIST_EMPTY = "Sorry, you have 0 MCs counted towards your GPA ._.\n"
             + "Please add some modules or grades!";
+
     public static final String ERROR_MODULAR_CREDITS_FAILED = "modular credits (accepted range: 0 to 100)";
```
Comment (Yzkkk) [inline src/main/java/seedu/duke/util/StringConstants.java]: The range was changed to 0 to 20 btw

Comment (heekit73098) [reply src/main/java/seedu/duke/util/StringConstants.java]: ok changed

File: src/main/java/seedu/duke/util/StringConstants.java (hunk 4 above)
Comment (heekit73098) [reply src/main/java/seedu/duke/util/StringConstants.java]: changed the format to include the invalid input as well

File: src/main/java/seedu/duke/util/StringConstants.java (hunk 5 above)
Comment (heekit73098) [reply src/main/java/seedu/duke/util/StringConstants.java]: same as above

File: src/main/java/seedu/duke/parsers/ModHappyParser.java (hunk 2 above)
Comment (heekit73098) [reply src/main/java/seedu/duke/parsers/ModHappyParser.java]: Shifted the logic to abstract class AddParser

File: src/main/java/seedu/duke/parsers/ModHappyParser.java (hunk 3 above)
Comment (heekit73098) [reply src/main/java/seedu/duke/parsers/ModHappyParser.java]: shifted the logic to abstract class EditParser

**Pull request merged**